### PR TITLE
[SPARK-36840][SQL] Support DPP if there is no selective predicate on the filtering side

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This pr makes it insert a DPP if there is no selective predicate on the filtering side and it still has benefits even makes the filter ratio much smaller.

### Why are the changes needed?

In some cases, it may pruning a lot of data even if there is no selective predicate on the filtering side.

Before this PR | After this PR
--- | ---
image  | image

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test.

### CODE REVIEW

**Comment by wangyum** [comment]
cc @cloud-fan  @maryannxue

File: sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/PartitionPruning.scala [hunk 1]
```diff
@@ -225,16 +230,6 @@ object PartitionPruning extends Rule[LogicalPlan] with PredicateHelper with Join
     }.isDefined
   }
 
-  /**
-   * To be able to prune partitions on a join key, the filtering side needs to
-   * meet the following requirements:
-   *   (1) it can not be a stream
-   *   (2) it needs to contain a selective predicate used for filtering
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/PartitionPruning.scala:232]: IIUC, the eventual goal is to make sure the "filtering side" is small. How do you guarantee it without selective predicates?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/PartitionPruning.scala:232]: In your example, the "filtering side" is a very small parquet table with 300+ rows. I don't think that's a general case and can justify removing the selective predicate check.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/PartitionPruning.scala:232]: e.g. we may broadcast a table with 1 million rows, and DPP may not be beneficial in that case.

**Comment by wangyum** [comment]
Another case from cluster:
image

File: sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/PartitionPruning.scala (hunk 1 above)
Comment (wangyum) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/PartitionPruning.scala:226]: Base on another case https://github.com/apache/spark/pull/34070#issuecomment-973893038. May be we can support DPP if filtering side rows less than 1000(`spark.sql.optimizer.dynamicPartitionPruning.filteringSideThreshold`).

**Pull request closed without merging**